library/network/src/modules/PortAliases.rb

```diff
@@ -141,175 +97,143 @@ def AllowedPortNameOrNumber
+        name, port, aliases = service.chomp.split(" ")
 
-        if !sport_name.nil?
-          service_aliases = Convert.convert(
-            Builtins.union(
-              service_aliases,
-              Builtins.splitstring(sport_name, " ")
-            ),
-            from: "list",
-            to:   "list <string>"
-          )
-        end
-        # service is a port name, any space isn't allowed
-      elsif IsAllowedPortName(port)
-        found_alias_port = Ops.get(@SERVICE_NAME_TO_PORT, port) do
-          LoadAndReturnNameToPort(port)
-        end
-        if !found_alias_port.nil?
-          service_aliases = Builtins.add(
-            service_aliases,
-            Builtins.tostring(found_alias_port)
-          )
+        port = port.to_i
+        aliases = [name, aliases&.split].flatten.compact
+        service = find_by_port(port)
 
-          # search for another port-name aliases when port-number found
-          service_aliases = Convert.convert(
-            Builtins.union(
-              service_aliases,
-              Builtins.splitstring(
-                Ops.get(@SERVICE_PORT_TO_NAME, found_alias_port) do
-                  LoadAndReturnPortToName(found_alias_port)
-                end,
-                " "
-              )
-            ),
-            from: "list",
-            to:   "list <string>"
-          )
-        end
-      else
-        if !Builtins.contains(@cache_not_allowed_ports, port)
-          @cache_not_allowed_ports = Builtins.add(
-            @cache_not_allowed_ports,
-            port
-          )
-          Builtins.y2error("Port name '%1' is not allowed", port)
+        if service
+          service.aliases |= aliases
         else
-          Builtins.y2debug("Port name '%1' is not allowed", port)
+          service = Service.new(port, aliases)
+          @services << service
         end
-        return [port]
       end
 
-      Builtins.toset(service_aliases)
+      @services
     end
 
-    # Function returns if the requested port-name is known port.
-    # Known port have an IANA alias.
+    # Returns services database after performing some cleanup
     #
-    # @param  string port-name
-    # @return  [Boolean] if is known
-    def IsKnownPortName(port_name)
-      # function returns the requested port and aliases if exists
-      return true if Ops.greater_than(Builtins.size(GetListOfServiceAliases(port_name)), 1)
-
-      false
+    # Basically, it manipulates the `getent services` output to discard duplicated lines after
+    # removing the protocol.
+    #
+    # Returned lines will look like
+    #
+    # EtherNet/IP-1         2222
+    # EtherNet-IP-2         44818
+    # rfb                   5900 vnc-server
+    #
+    # @return [Array<String>]
+    def load_services_database
+      Yast::Execute.stdout.on_target!(
+        ["/usr/bin/getent", "services"],
+        # remove the protocol from the output
+        ["/usr/bin/sed", "-r", "s/(\\S)(\\/\\S+)(.*)?/\\1\\3/"],
```

Comment:
I know this looks a bit complicated. The intention here, as stated, is to filter out the _duplicated_ lines, where duplicate means "lines for the same service name, port and aliases but different protocol". I think this is faster than processing them in Ruby, but I'm not against to change it.

What do you think?

Reply:
It looks quite unreadable to me :wink:. I would give it a try using Ruby as I do not expect a noticeable performance impact.